[15.0][FIX] mail_show_follower: Use company from the document to render messages correctly

carlos-lopez-tecnativa opened the pull request:
In a multi-company environment, if one company uses the `show_followers_partner_format` by `partner_name` and another company uses `partner_email`, a user with permissions in both companies may experience issues. If a document belongs to Company A, but the current company is set to Company B, the message format is taken from the current company (Company B) instead of the document’s company (Company A). This fix ensures that the message format is rendered based on the document's company.

TT51022 @Tecnativa
@pedrobaeza @chienandalu could you please review it, thanks

vvrossem committed:
[16.0][FIX][REF] mail_show_follower: internal notes

This module aims to show the recipients of an email in the email header, but the previous implementation was not suitable for internal notes.

When **logging an internal note** with tagged users and/or partners, those specific recipients should be displayed
as well as the document's followers subscribed to this message subtype.

When **sending an email**, the recipients should be the document's followers.

This is achieved by using the union of `mail.mail`'s `recipient_ids` and `notified_partner_ids`.

pedrobaeza commented on mail_show_follower/models/mail_mail.py:
```diff
@@ -104,8 +104,10 @@ def _send(self, auto_commit=False, raise_exception=False, smtp_session=None):
                         # get partners
                         cc_internal = True
                         # else get company in object
+                        company = self.env.company
```
I think you can refactor this code to get the put the logic to get the company on one side, and then access to the properties then, so it's clearer. If not, all this `if` are hard to read and prone to errors.

carlos-lopez-tecnativa replied:
@pedrobaeza I backported the changes from version 16 because it includes a refactor that makes the code more readable. I squashed two commits into a single one and added a commit to fix the multi-company issue. Please let me know if you have any questions.

carlos-lopez-tecnativa committed:
[FIX] mail_show_follower: Use company from the document to render messages correctly + proper lang handle

In a multi-company environment, if one company uses the
show_followers_partner_format by partner_name and another company
uses partner_email, a user with permissions in both companies may
experience issues. If a document belongs to Company A, but the current
company is set to Company B, the message format is taken from the
current company (Company B) instead of the document’s company
(Company A). This fix ensures that the message format is rendered based
on the document's company.

It also fixes the lang handling for cases with False values.

carlos-lopez-tecnativa commented:
Backport of https://github.com/OCA/social/pull/1391

yajo commented on mail_show_follower/models/mail_mail.py:
```diff
@@ -59,97 +59,31 @@ def remove_p(markup_txt):
+            # recipients from any Notification Type (i.e. email, inbox, etc.)
+            recipients = mail.notification_ids.res_partner_id
+            record = self.env[mail.model].browse(mail.res_id)
+            company = getattr(record, "company_id", self.env.company)
```
suggestion: This is not the recommended way to get a field that you don't know if exists, because that attribute could perfectly be a method (imagine `def company_id():`) and break your logic. Hardly probable, but possible.


The rule of thumb way to make sure it's really a field:

```suggestion
            company = record["company_id"]
```

Of course there can be a `KeyError` which you can circumvent with a try-except block, or a `if "company_id" in record` guard.

yajo commented on mail_show_follower/models/mail_mail.py:
```diff
@@ -59,97 +59,31 @@ def remove_p(markup_txt):
+                or company.partner_id.lang
+            )
+            final_cc = (
+                mail.with_context(lang=langs[0])
```
issue: `langs` is of type `list[str] | str | bool` as per the code above. This is only supporting the `list[str]` case.

suggestion: I think the bug you guys discovered deserves at least a new test.

yajo commented on mail_show_follower/models/mail_mail.py:
```diff
@@ -59,97 +59,31 @@ def remove_p(markup_txt):
+            recipients = mail.notification_ids.res_partner_id
+            record = self.env[mail.model].browse(mail.res_id)
+            company = getattr(record, "company_id", self.env.company)
+            show_internal_users = company and company.show_internal_users_cc
```
suggestion: This guard is unnecessary:

```suggestion
            show_internal_users = company.show_internal_users_cc
```

yajo commented on mail_show_follower/models/mail_mail.py:
```diff
@@ -59,97 +59,31 @@ def remove_p(markup_txt):
-                        # it is saved in the body_html field so that it does
-                        # not appear in the odoo log
-                        mail.body_html = final_cc + mail.body_html
+        group_user = self.env.ref("base.group_user")
```
issue: although the PR description explains a clear and concise problem, the code I see here is a medium refactor of a very important (and yes, spaghethic) piece of code.

For example, you are going away from testing portal group membership into testing backend user membership.

So, the PR goes beyond scope and gets harder to review. Could you split the fix and the refactor in 2 different PRs? Or could you explain a little bit more what's this refactor doing, so it's easier to review?

yajo reviewed (changes requested):
Good job, thanks! Some concerns below.

pedrobaeza commented:
@yajo this is a backport of #1391, with only an extra patch in a separated commit, so all your comments should go to the 16.0 PR when it was done, and it was you who approve it. We are not going to invest more time in that part, as we will finally get to the newest version with the refactoring already done.

The only patch from us applied right now is the company one, but we have changed also the lang part, which indeed can have some side effects (but this is not new). We will fw-port it to 16 if this PR is accepted.

Please accept it as it is to get both versions with the good refactoring, or we will let it die as a PR injected where needed.

yajo reviewed (approved):
Thanks! I didn't realize this was a backport and there were 2 commits. Also thanks for that extra fix. Happy to help! ❤️ 

/ocabot merge minor

pedrobaeza commented:
@carlos-lopez-tecnativa please fw-port the second patch to 16
